Add support for reverse consumer information

mook-as opened the pull request:
In `/var/vcap/job-src/*/config_spec.json`, we will now include a list of consumers of this job's links (via the `consumed_by` key, similar to the existing `consumes` key).  This is useful so we can attempt to restart pods which contain outdated information (in case they linked against a previous iteration of this pod and now contains outdated information).

/cc @viovanov

mook-as committed:
Add support for reverse consumer information

In /var/vcap/job-src/*/config_spec.json, we will now include a list of
consumers of this job's links (via the `consumed_by` key, similar to the
existing `consumes` key).  This is useful so we can attempt to restart
pods which contain outdated information (in case they linked against a
previous iteration of this pod and now contains outdated information).

f0rmiga commented on model/job_reference.go:
```diff
@@ -11,7 +11,8 @@ type JobReference struct {
 	ReleaseName         string                     `yaml:"release"` // The release the job comes from
 	ExportedProviders   map[string]JobProvidesInfo `yaml:"provides"`
-	ResolvedConsumers   map[string]JobConsumesInfo `yaml:"consumes"`
+	ResolvedConsumers   map[string]JobConsumesInfo `yaml:"consumes"`    // Instance groups that this job links to & requires
```
How about aligning the struct field label with the yaml field label? It is a bit confusing one being `...Consumers` and the other `consumes`.

mook-as replied:
I feel like that should be done separately (this was just a comment change; refactoring would be a bigger change that should get its own PR).  I might still be convinced though!

f0rmiga replied:
It is ok to do in a separate PR, so we don't mix things up! Could we do right after?

f0rmiga replied:
Card created to tackle this.

f0rmiga commented on model/resolver/resolver.go:
```diff
@@ -345,6 +345,43 @@ func (r *Resolver) ResolveLinks() validation.ErrorList {
 		}
 	}
 
+	// Record in each job what roles consume it
```
[nit] Punctuation here and below.

f0rmiga commented on model/resolver/resolver.go:
```diff
@@ -345,6 +345,43 @@ func (r *Resolver) ResolveLinks() validation.ErrorList {
 	}
 
+	// Record in each job what roles consume it
+	for _, consumerInstanceGroup := range m.InstanceGroups {
```
What do you think about pulling the logic from the inner-most loop into a function?

f0rmiga commented on model/resolver/resolver.go:
```diff
@@ -345,6 +345,43 @@ func (r *Resolver) ResolveLinks() validation.ErrorList {
+					// This should not happen: we resolved a link, but can no
+					// longer find the instance group that provides it
+					errors = append(errors, validation.InternalError(
+						fmt.Sprintf(`instance_group[%s].job[%s].consumes[%s]`, consumerInstanceGroup.Name, consumerJob.Name, linkName),
```
Declaring variables in a more procedural approach would help reading it. E.g.:
```go
field := fmt.Sprintf("instance_group[%s].job[%s].consumes[%s]", consumerInstanceGroup.Name, consumerJob.Name, linkName)
err := fmt.Errorf("Could not find resolved instance group %s", consumer.RoleName)
internalErr := validation.InternalError(field, err)
errors = append(errors, internalErr)
```

Also, no need for using ` for the strings.

f0rmiga commented on model/resolver/resolver.go:
```diff
@@ -345,6 +345,43 @@ func (r *Resolver) ResolveLinks() validation.ErrorList {
+						fmt.Errorf(`Could not find resolved job %s in instance group %s`, consumer.JobName, consumer.RoleName)))
+					continue
+				}
+				if providerJob.ResolvedConsumedBy == nil {
```
I think it should be initialized when the object is created.

f0rmiga commented on model/resolver/validation.go:
```diff
@@ -54,6 +54,10 @@ func validateInstanceGroup(roleManifest *model.RoleManifest, g *model.InstanceGr
 			jobReference.ResolvedConsumers[name] = info
 		}
+
+		if jobReference.ResolvedConsumedBy == nil {
```
Why a validation function is mutating a parameter passed to it? I see the previous code is doing it already, but it is way out of its concern doing so. Could you add a TODO for moving it out, and create a card for it on our board?

f0rmiga replied:
Added a card to tackle this.

mook-as committed:
Address review comments, stage 1

https://github.com/cloudfoundry-incubator/fissile/pull/470#pullrequestreview-223030846

This is just the trivial changes.

mook-as committed:
model/resolve: move mostly-independent code to a function

This lets us keep less state in our brain as we read code.

https://github.com/cloudfoundry-incubator/fissile/pull/470/files/0d5294f5a9d269319f4c69de9a56ad551dc7cd10#r272394050
